### PR TITLE
[M68k] Add remaining addressing modes for Atomic operations

### PR DESCRIPTION
**Pull request by knickish**
Had been doing this piece by piece, but makes more sense to do it in a single PR. Adds support for `ARID`, `PCI`, `PCD`, `AL`, and `ARD` addressing modes for atomic operations, along with a variety of tests. 

The `CodeModel` tests have been rearranged, as some of the new addressing modes are only exercised under some combinations of `CodeModel` and relocation mode

### CODE REVIEW

**Comment by glaubitz** [comment]
Thanks. I'm going to test this and report back. Sorry for not being able to test your previous patch, I was busy with work.

**Comment by knickish** [comment]
Thanks, am curious to hear if it works as expected

**Comment by glaubitz** [comment]
The backend still seems to have problems to select certain atomic instructions:

```
rustc-LLVM ERROR: Cannot select: t14: ch = AtomicStore<(store unordered (s8) into %ir.25)> t13:1, t13, t9
  t13: i8,ch = AtomicLoad<(load unordered (s8) from %ir.26)> t0, t12
    t12: i32 = add t11, t2
      t11: i32,ch = CopyFromReg t0, Register:i32 %0
        t10: i32 = Register %0
      t2: i32,ch = CopyFromReg t0, Register:i32 %8
        t1: i32 = Register %8
  t9: i32 = add t8, t2
    t8: i32,ch = CopyFromReg t0, Register:i32 %1
      t7: i32 = Register %1
    t2: i32,ch = CopyFromReg t0, Register:i32 %8
      t1: i32 = Register %8
In function: __llvm_memmove_element_unordered_atomic_1
error: rustc interrupted by SIGSEGV, printing backtrace
```

**Comment by glaubitz** [comment]
I tried building rustc from git master, but that fails because now rust will always try to build the embedded copy of LLVM.

**Comment by glaubitz** [comment]
OK, I managed to get further and noticed that I also needed #114714. After that, it already came to the linker stage, so that's a huge progress.

At that point, I needed to add M68k support to the object crate only to find out you had just done that. :-D I'll keep on testing.

Let's hope that @mshockwave and @0x59616e can quickly review both this PR and #114714 because it would be the first time we'd be able to build libstd.

**Comment by glaubitz** [comment]
@knickish Can you tell me what changes are necessary for Rust to generate EM_68K binaries?

I tried your patch on top of the `object` crate plus the following change:

```
diff --git a/compiler/rustc_codegen_ssa/src/back/metadata.rs b/compiler/rustc_codegen_ssa/src/back/metadata.rs
index 3f3cb8b4073..c50a5eb5d64 100644
--- a/compiler/rustc_codegen_ssa/src/back/metadata.rs
+++ b/compiler/rustc_codegen_ssa/src/back/metadata.rs
@@ -197,6 +197,7 @@ pub(crate) fn create_object_file(sess: &Session) -> Option<write::Object<'static
         ),
         "x86" => (Architecture::I386, None),
         "s390x" => (Architecture::S390x, None),
+        "m68k" => (Architecture::M68k, None),
         "mips" | "mips32r6" => (Architecture::Mips, None),
         "mips64" | "mips64r6" => (Architecture::Mips64, None),
         "x86_64" => (
```

but Rust still uses the generic ELF target for m68k.

**Comment by glaubitz** [comment]
Ah, I think I need to build a bootstrap compiler first which includes the change to `rustc_codegen_ssa` as otherwise it won't be possible to build any artifacts for m68k.

**Comment by glaubitz** [comment]
Hmm, building a bootstrap compiler with the patch didn't help.

I had this problem in the past when adding support for 32-bit SPARC to the `object` crate, but I absolutely cannot remember how I fixed it.

**Comment by knickish** [comment]
> @knickish Can you tell me what changes are necessary for Rust to generate EM_68K binaries?

@ glaubitz I cannot, I am stuck on this also, at least for large enough binaries :(  Have been experimenting with the relocation model linker flags and such, but can't seem to get it working. I will ping you on rust zulip to continue discussing if that works for you

File: llvm/lib/Target/M68k/M68kISelDAGToDAG.cpp [hunk 1]
```diff
@@ -708,6 +708,20 @@ bool M68kDAGToDAGISel::SelectARIPD(SDNode *Parent, SDValue N, SDValue &Base) {
   return false;
 }
 
+static bool allowARIDWithDisp(SDNode *Parent) {
```
Comment (mshockwave) [inline llvm/lib/Target/M68k/M68kISelDAGToDAG.cpp:711]: will the compiler prompt unused function warning when building LLVM without assertion?

File: llvm/lib/Target/M68k/M68kInstrAtomics.td [hunk 2]
```diff
@@ -36,17 +57,128 @@ class MxCASOp<bits<2> size_encoding, MxType type>
...
+  let Constraints = "$out = $dc";
+  let mayLoad = 1;
+  let mayStore = 1;
+}
+
+def CASARID8  : MxCASARIDOp<0x1, MxType8d>;
+def CASARID16 : MxCASARIDOp<0x2, MxType16d>;
+def CASARID32 : MxCASARIDOp<0x3, MxType32d>;
+
+class MxCASARIIOp<bits<2> size_encoding, MxType type>
```
Comment (mshockwave) [inline llvm/lib/Target/M68k/M68kInstrAtomics.td:81]: Do we have MC tests for these new instructions (I assume you already have their codegen tests)? If not, could you add them?

Ditto for other CAS variants below.

**Review by mshockwave — COMMENTED** [review]
Thanks for the patch! I only have some comments regarding MC tests

**Comment by glaubitz** [comment]
@knickish Could you begin the commit summary with a capital letter and maybe replace "remaining" with "missing"? And I think "pic" should be all all-caps, i.e. "PIC" as it's an acronym.

As for the testing goes, I can confirm that this change fixes the remaining LLVM backend errors we have seen so far when building Rust code on M68k which is amazing ;-).

File: llvm/lib/Target/M68k/M68kInstrAtomics.td (hunk 2 above)
Comment (knickish) [reply llvm/lib/Target/M68k/M68kInstrAtomics.td:81]: Added a new test at `llvm/test/MC/M68k/Atomics/cas.s` which shows all of them except `ARD`, which I couldn't get to parse for some reason. I was getting an illegal operand error when I tried to add `cas.w %d4, %d5, %a3` as a test case. Am I expecting something to work incorrectly here, or is it likely an issue with the AsmParser?

File: llvm/lib/Target/M68k/M68kISelDAGToDAG.cpp (hunk 1 above)
Comment (knickish) [reply llvm/lib/Target/M68k/M68kISelDAGToDAG.cpp]: style guide was unclear about what do in this case, so I added `[[maybe_unused]]` here

File: llvm/lib/Target/M68k/M68kInstrAtomics.td (hunk 2 above)
Comment (knickish) [reply llvm/lib/Target/M68k/M68kInstrAtomics.td:81]: On further investigation, looks like PC-relative and `ARD` addressing modes are not legal for `CAS`, will have to lower these to some different instruction sequences

Comment (knickish) [reply llvm/lib/Target/M68k/M68kInstrAtomics.td:81]: Removed the offending modes for now, as it looks like the code changes will be in different files. All of the addressing modes added by the patch are now tested in the new `MC` test

**Comment by knickish** [comment]
Ping @mshockwave I think everything is addressed that you mentioned before

**Review by mshockwave — APPROVED** [review]
LGTM thanks!

**Comment by glaubitz** [comment]
@mshockwave Could you land this change as well?

**Pull request merged**